service/lib/dinstaller/config_reader.rb

```diff
@@ -0,0 +1,170 @@
+# under the terms of version 2 of the GNU General Public License as published
+# by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+# more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, contact SUSE LLC.
+#
+# To contact SUSE LLC about this file by physical or electronic mail, you may
+# find current contact information at www.suse.com.
+
+require "yast"
+require "logger"
+require "dinstaller/config"
+require "dinstaller/cmdline_args"
+require "transfer/file_from_url"
+
+Yast.import "URL"
+Yast.import "Directory"
+
+module DInstaller
+  # This class is responsible for reading DInstaller configuration from different locations
+  # including kernel cmdline options
+  class ConfigReader
+    include Yast::Transfer::FileFromUrl
+    include Yast::I18n
+
+    # Default DInstaller configuration wich should define all the possible values
+    SYSTEM_PATH = "/etc/d-installer.yaml"
+    GIT_PATH = File.expand_path("#{__dir__}/../../etc/d-installer.yaml")
+    REMOTE_BOOT_CONFIG = "d-installer_boot.yaml"
+
+    PATHS = [
+      "/usr/lib/d-installer.d",
+      "/etc/d-installer.d",
+      "/run/d-installer.d"
+    ].freeze
+
+    attr_reader :logger
+    attr_reader :workdir
+
+    # Constructor
+    #
+    # @param logger [Logger]
+    # @param workdir [String] Root directory to read the configuration from
+    def initialize(logger: nil, workdir: "/")
+      @logger = logger || ::Logger.new($stdout)
+      @workdir = workdir
+    end
+
+    # Return an {Array} with the different {Config} objects read from the different locations
+    #
+    # TODO: handle precedence correctly
+    #
+    # @returm [Array<Config>] an array with all the configurations read from the system
+    def configs
+      return @configs if @configs
+
+      @configs = config_paths.map { |path| config_from_file(path) }
+      @configs << remote_config if remote_config
+      @configs << cmdline_config if cmdline_config
+      @configs
+    end
+
+    # Return a {Config} oject
+    # @return [Config] resultant Config after merging all the configurations
+    def config
+      config = configs.first || Config.new
+      (configs[1..-1] || []).each { |c| config = config.merge(c) }
+      config
+    end
+
+  private
+
+    # Copy a file from a potentially remote location
+    #
+    # @param location [String] File location. It might be an URL-like string (e.g.,
+    #   "http://example.net/example.yml").
+    # @param target [String] Path to copy the file to.
+    # @return [Boolean] Whether the file was sucessfully copied or not
+    def copy_file(location, target)
+      url = Yast::URL.Parse(location)
+
+      res = get_file_from_url(
```

Comment:
NP: I'd use _result_ or something different to _res_ :) :)

Reply:
Ok, maybe next time I touch it if there is not a big change requested by now... ;) Thanks